File: day22/.ts
function canEscape(maze: string[][]): boolean {
  const startRow = maze.findIndex(row => row.includes("S"))
  let startPoint = [
    startRow,
    maze[startRow].indexOf("S")
  ]
  const visited = Array.from({ length: maze.length }, () => Array(maze[0].length).fill(false));

  const dirs = [
    [1, 0],
    [-1, 0],
    [0, 1],
    [0, -1]
  ]

  function isValid(i, j) {
    return i >= 0 && i < maze.length && j >= 0 && j < maze[0].length && maze[i][j] != "#"
  }

  function dfs(i, j){
    if (maze[i][j] == "E") return true
    if (visited[i][j]) return false

    visited[i][j] = true

    for (const [x, y] of dirs){
      const ni = i + x;
      const nj = j + y;
      if (isValid(ni, nj)) {
        if (dfs(ni, nj)) return true;
      }
    }
    return false
  }
  const result = dfs(startPoint[0], startPoint[1])
  return result
}

File: day22/.py
from collections import deque

def can_escape(maze: list[list[str]]) -> bool:
  rows, cols = len(maze), len(maze[0])
  visited = [[False] * cols for _ in range(rows)]

  for i in range(rows):
    for j in range(cols):
      if maze[i][j] == "S":
        start_i, start_j = i, j
        break

  dirs = [(1,0), (-1,0), (0,1), (0,-1)]
  queue = deque([(start_i, start_j)])
  visited[start_i][start_j] = True

  while queue:
    i, j = queue.popleft()

    if maze[i][j] == "E":
      return True

    for di, dj in dirs:
      ni, nj = i + di, j + dj
      if 0 <= ni < rows and 0 <= nj < cols and maze[ni][nj] != "#" and not visited[ni][nj]:
        visited[ni][nj] = True
        queue.append((ni, nj))

  return False
